### PR TITLE
server: fix dead lock problem in raft cluster

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve? 
After merging #1654, there are some lock bugs which will result in deadlock problems sometimes.

### What is changed and how it works?
This PR is going to fix them.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by disksing** [comment]
Please take a look @Luffbee

File: server/cluster.go [hunk 1]
```diff
@@ -1169,8 +1169,8 @@ func (c *RaftCluster) AllocPeer(storeID uint64) (*metapb.Peer, error) {
 
 // OnStoreVersionChange changes the version of the cluster when needed.
 func (c *RaftCluster) OnStoreVersionChange() {
-	c.Lock()
-	defer c.Unlock()
+	c.RLock()
+	defer c.RUnlock()
```
Comment (Luffbee) [inline server/cluster.go]: Is it really safe? I see two functions have write semantic: `c.opt.SetClusterVersion()` and `c.opt.Persist()`. And there is no code or comments indicating they are thread-safe.

File: server/grpc_service.go [hunk 2]
```diff
@@ -627,8 +625,8 @@ func (s *Server) ScatterRegion(ctx context.Context, request *pdpb.ScatterRegionR
 	}
 
 	cluster.RLock()
-	defer cluster.RUnlock()
 	co := cluster.coordinator
+	cluster.RUnlock()
```
Comment (Luffbee) [inline server/grpc_service.go]: This `RLock()` is like the above one, no meaning.

File: server/cluster.go [hunk 3]
```diff
@@ -201,23 +201,30 @@ func (c *RaftCluster) loadClusterInfo() (*RaftCluster, error) {
...
-		zap.Int("count", c.core.Stores.GetStoreCount()),
+		zap.Int("count", c.getStoreCount()),
 		zap.Duration("cost", time.Since(start)),
 	)
 
 	start = time.Now()
+
+	c.core.Lock()
 	if err := c.storage.LoadRegions(c.core.Regions); err != nil {
+		c.core.Unlock()
```
Comment (nolouch) [inline server/cluster.go]: Can we let reload regions to `core` package?

**Comment by rleungx** [comment]
@nolouch PTAL.

**Pull request merged**